presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HistoricalStatisticsEquivalentPlanMarkingOptimizer.java

```diff
@@ -93,11 +102,32 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Var
         }
 
         // Assign 'statsEquivalentPlanNode' to plan nodes
-        plan = SimplePlanRewriter.rewriteWith(new Rewriter(idAllocator), plan, new Context());
+        PlanNode newPlan = SimplePlanRewriter.rewriteWith(new Rewriter(idAllocator), plan, new Context());
+        // Return original plan if timeout
+        if (checkTimeOut(startTimeInNano, timeoutInMilliseconds)) {
+            logOptimizerFailure(session);
+            return plan;
+        }
 
         // Fetch and cache history based statistics of all plan nodes, so no serial network calls happen later.
-        statsCalculator.registerPlan(plan, session);
-        return plan;
+        boolean registerSucceed = statsCalculator.registerPlan(newPlan, session, startTimeInNano, timeoutInMilliseconds);
+        // Return original plan if timeout or registration not successful
+        if (checkTimeOut(startTimeInNano, timeoutInMilliseconds) || !registerSucceed) {
+            logOptimizerFailure(session);
+            return plan;
+        }
+        return newPlan;
+    }
+
+    private boolean checkTimeOut(long startTimeInNano, long timeoutInMilliseconds)
+    {
+        return NANOSECONDS.toMillis(System.nanoTime() - startTimeInNano) > timeoutInMilliseconds;
+    }
+
+    private void logOptimizerFailure(Session session)
+    {
+        session.getWarningCollector().add((new PrestoWarning(QUERY_OPTIMIZER_FAILURE, format("'%s' failed to work", HistoricalStatisticsEquivalentPlanMarkingOptimizer.class.getSimpleName()))));
```

Comment:
do we want to log additional information about the failure?
do we plan to use this for other usecases beyond HBO timeout?

Reply:
perhaps make the function more general and pass in the optimizer name as argument

Reply:
> do we want to log additional information about the failure?

Maybe just this for now, we can add more later if needed.

> do we plan to use this for other usecases beyond HBO timeout?

Other optimizers can populate this field as well.

> perhaps make the function more general and pass in the optimizer name as argument

This is a private method in HBO optimizer, will not be used by other optimizers